### ISSUE TITLE
PDF Scanner Misses Most / All Emails

### ISSUE THREAD

**Issue by mthbrown** [simsong/bulk_extractor#373]
Hi,

I was just testing out **bulk_extractor**. One of my tests was to create the following text file:

```
abc@google.com
def@google.com
def@gmail.com
abc@google.com
abc@google.com
```

When I then run bulk_extractor and point it to the text file, I get the expected in `email.txt`:

```
# BANNER FILE NOT PROVIDED (-b option)
# BULK_EXTRACTOR-Version: 2.0.0
# Feature-Recorder: email
# Filename: /tmp/1.txt
# Feature-File-Version: 1.1
0       abc@google.com  abc@google.com\015\012def@google.com
16      def@google.com  abc@google.com\015\012def@google.com\015\012def@gmail.com\015
32      def@gmail.com   def@google.com\015\012def@gmail.com\015\012abc@google.com
47      abc@google.com  \012def@gmail.com\015\012abc@google.com\015\012abc@google.com
63      abc@google.com  abc@google.com\015\012abc@google.com\015\012
```

I then converted the text file to a PDF using **pdflatex** and opened it in a PDF file and I can clearly see the PDFs (on a single line with spaces between them) as shown here:

![Screenshot from 2022-09-19 07-46-43](https://user-images.githubusercontent.com/22183215/190929273-961da02e-5846-44f1-b695-8ddc31b5bfb7.png)

Now I only get this when I run bulk_extractor:

```
# BANNER FILE NOT PROVIDED (-b option)
# BULK_EXTRACTOR-Version: 2.0.0
# Feature-Recorder: email
# Filename: emails.pdf
# Feature-File-Version: 1.1
69-PDF-35       def@gmail.com   e f@go ogle.com def@gmail.com ab c@go ogle.co
```

Finally, when I then opened the text file in Firefox and selected **Print to PDF** and opened the file in a PDF reader and it showed me the expected text:

![Screenshot from 2022-09-19 07-48-46](https://user-images.githubusercontent.com/22183215/190929328-2eeed0bb-5d51-4c0b-a9f4-b6f2458a13fb.png)

However, now when I run bulk_extractor on the generated PDF, `email.txt` is empty. Is this expected behavior? Am I missing something? Thanks

**Comment by simsong** [comment]
Hi. Thank you for submitting the bug report. Would it be possible for you to attach the two PDFs to this ticket?

It turns out that the `bulk_extractor` PDF to text program does not work the way that most PDF to text programs work, as it is designed to work with fragmented files. Instead of going to the end of the PDF file, reading a table, going to each page, creating the objects, and then interpreting the objects, [scan_pdf](https://github.com/simsong/bulk_extractor/blob/main/src/scan_pdf.cpp) looks for patterns within the inflated compressed streams and applies some simple heuristics. The heuristics were based on analysis of PDF files in the 2008-2014 time period. But the way the PDFs are created from text changes over time. `bulk_extractor` was not designed for `pdflatex` or for `Firefox` PDF generators. It was designed for Microsoft Word on the Mac and Windows.

A better heuristic would be to take all of the (x,y) locations of the text, drop them into a frame buffer, and then run OCR on the frame buffer. You wouldn't need to do full OCR because you already know what the letters are. You would need to d line and word break detection.  You need to find the lines so you know the order to send the characters, and you need the word break because there are no spaces encoded in PDF files.

Do you want to give this a try? `bulk_extractor` has switches to dump the inflated compressed streams, and then you can write new recognizer that turns the characters into a text stream.

**Comment by mthbrown** [comment]
Thanks @simsong. I added the PDFs. Unfortunately, I don't know C++

**Comment by simsong** [comment]
This is an easy way Iearn!

Do you know python? I have been planning on doing a python bridge


----
Sent from my phone.

On Sep 18, 2022, at 8:21 PM, mthbrown ***@***.***> wrote:

﻿

Thanks @simsong<https://github.com/simsong>. I added the PDFs. Unfortunately, I don't know C++

**Comment by mthbrown** [comment]
I know some Python

**Comment by q2dg** [comment]
I don't know if this is related or not but doing PDF scanner over any random PDF file doesn't show any string shown with _strings_ command flawessly.

**Comment by simsong** [comment]
> I don't know if this is related or not but doing PDF scanner over any random PDF file doesn't show any string shown with _strings_ command flawessly.

Hi. What do you mean? I ran `strings` over your two PDFs and there was no output.

**Comment by simsong** [comment]
Those strings have nothing to do with the PDF’s visible content. Why would you want to report strings from PDF internal structures? That’s not useful.
There are no emails in that data.

**Comment by q2dg** [comment]
Ok, thanks! 
So, just for clarification: does pdf scanner scans any textual (visible) content of a pdf or just email? Description by bulk_scanner -H says "Extract text from PDF files". If so, why the word "Hola" isn't extracted from my pdf file when running any bulk_extractor command I put in my former message?
Thanks again

**Comment by q2dg** [comment]
Yes, I'm a bulk_extractor user. But I recognize I did't understand the pdf scanner behaviour. It's my fault, sorry

**Comment by simsong** [comment]
> Yes, I'm a bulk_extractor user. But I recognize I did't understand the pdf scanner behaviour. It's my fault, sorry

It's not your fault, it's bad documentation!  This PDF scanner is designed to extract data from fragments of PDF files on disks by extracting from within the PDF the characters that are printed onto a page. This is different from the way that PDFs are rendered. However, all of the data within the PDF (metadata and printed data) are scanned for things like credit card numbers and email addresses. But bulk_extractor doesn't dump the text that it extracts, it just passes it to the next parser.

You might want to read the bulk_extractor paper for more details:

https://www.sciencedirect.com/science/article/pii/S0167404812001472

**Comment by simsong** [comment]
Here's the decompressed stream from pandoc.doc:
```
%PDF-1.5
%<E2><E3><CF><D3>
1 0 obj

<<
/S /GoTo
/D [2 0 R /Fit]
>>
endobj

2 0 obj

<<
/pdftk_PageNum 1
/Contents 3 0 R
/Type /Page
/Resources 4 0 R
/Parent 5 0 R
/MediaBox [0 0 612 792]
>>
endobj

4 0 obj

<<
/ProcSet [/PDF /Text]
/Font
<<
/F28 6 0 R
>>
>>
endobj

3 0 obj

<<
/Length 249
>>
stream
0 g 0 G
0 g 0 G
0 g 0 G
BT
/F28 9.9626 Tf 133.768 657.235 Td [(ab)-28(c@go)-28(ogl)1(e.com)-223(de)1(f@go)-28(ogle.com)-222(def@gmail.com)-222(ab)-28(c@go)-28(ogle.com)-222(ab)-28(c@go)-27(ogle.com)]TJ
0 g 0 G
 169.365 -567.87 Td [(1)]TJ
0 g 0 G
ET

endstream
endobj
```

Generated with this command line:
```
pdftk pandoc.pdf output pandoc-uncompressed.pdf uncompress
```
